gnupg-pkcs11-scd/keyutil.c

```diff
@@ -118,43 +217,87 @@ keyutil_get_cert_mpi (
 		error = GPG_ERR_BAD_CERT;
 		goto cleanup;
 	}
- 
+
 	if ((pubkey = X509_get_pubkey (x509)) == NULL) {
 		error = GPG_ERR_BAD_CERT;
 		goto cleanup;
 	}
- 
-	if ((rsa = EVP_PKEY_get1_RSA(pubkey)) == NULL) {
-		error = GPG_ERR_WRONG_PUBKEY_ALGO;
+
+	pubkey_ctx = EVP_PKEY_CTX_new(pubkey, NULL);
+	if (pubkey_ctx == NULL) {
+		error = GPG_ERR_BAD_CERT;
 		goto cleanup;
 	}
 
-	RSA_get0_key(rsa, &n, &e, NULL);
-
-	n_hex = BN_bn2hex (n);
-	e_hex = BN_bn2hex (e);
-
-	if(n_hex == NULL || e_hex == NULL) {
-		error = GPG_ERR_BAD_KEY;
+	/**
+	 * Check the public key context
+	 * 1 is success, -2 is not applicable
+	 */
+	check_result = EVP_PKEY_public_check(pubkey_ctx);
+	if (check_result != 1 && check_result != -2) {
+		error = GPG_ERR_WRONG_PUBKEY_ALGO;
```

Comment:
this is interesting, are you sure we want to continue if the public algorithm is not supported? I understand we use the PKCS\#11 to perform the operation, but still it should be supported... and better to have a definite success.

Reply:
I'm not sure what the comment means -- but if we're not able to identify the kind of key we won't be able to figure out how to populate `keyinfo` so we won't be able to get to the point of signing because we wouldn't know how to format all of the supporting pieces up- and down-stream (up to the agent, how would we indicate what kind of key it is?;  down to the PKCS#11 module, which mechanism would we say we wanted to use ?)

Reply:
my question was why do you accept result `-2`, why don't you allow only success.